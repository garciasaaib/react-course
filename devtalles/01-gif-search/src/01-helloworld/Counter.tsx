import { useState } from "react";

export const Counter = ({initialValue = 0}) => {
  const [counter, setCounter] = useState(initialValue);
  const handleAdd = () => {
    setCounter((currValue) => currValue+1);
  }
  const handleSubstract = () => {
    setCounter((currValue) => currValue-1);
  }
  const handleReset = () => {
    setCounter(0);
  }
  return (
    <div>
      <h2>Counter: {counter}</h2>
      <button onClick={handleAdd}>Add</button>
      <button onClick={handleReset}>Reset</button>
      <button onClick={handleSubstract}>Substract</button>

    </div>
  )
}
