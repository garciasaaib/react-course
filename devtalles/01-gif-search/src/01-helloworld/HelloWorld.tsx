import { Counter } from "./Counter";

// un componente padre
export const HelloWorld = ({ title = "holamundo" }) => {
	return (
		<section>
			<Title title={title} />
			<Counter initialValue={1} />
		</section>
	);
};

// un componente hijo
export const Title = ({ title = "hola soy goku" }: { title?: string }) => (
	<h1
		style={{ background: "teal" }}
		className="custom inline"
		data-testid="test-title"
	>
		{title}
	</h1>
);

/** Notas
 * - <></> es un fragmento de React, es como un div pero no se renderiza en el DOM
 * - Se puede usar React.Fragment o <>
 * - Las variables y funciones estaticas pueden vivir preferentemente fuera del componente
 * - Los styles se agregan como un objeto style al componente se agregan con className o con un objeto
 * - props son un objeto que permiten la comunicacion entre componente padre e hijo
 * - DefaultProps es una propiedad que permite definir valores por defecto a las props, eso las hace opcionales
 * - con proptypes puedes definir el tipo de dato que esperas recibir, pero si tienes ts no necesitas proptypes
 */
