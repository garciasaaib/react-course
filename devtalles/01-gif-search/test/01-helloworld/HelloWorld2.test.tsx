import React from "react";
import { render, screen } from "@testing-library/react";
import { HelloWorld } from "../../src/01-helloworld/HelloWorld";

describe("<HelloWorld />", () => {
	// inicializacion de variables
	const title = "holoroco";

	test("Debe hacer match con el snapshoot", () => {
		const { container } = render(<HelloWorld title={title} />);
		expect(container).toMatchSnapshot();
	});

	test("Debe mostrar el mensaje con title", () => {
		render(<HelloWorld title={title} />);

		// nos hace una copia simplificada del dom para poder hacer pruebas
		// screen.debug();

		expect(screen.getByText(title)).toBeInTheDocument();
	});

	test("Debe mostrar el mensaje con title en un h1", () => {
		render(<HelloWorld title={title} />);
		expect(screen.getByRole("heading", { level: 1 }).innerHTML).toContain(title);

	});

});
