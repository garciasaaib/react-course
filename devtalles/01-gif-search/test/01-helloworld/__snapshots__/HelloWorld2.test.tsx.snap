// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelloWorld /> Debe hacer match con el snapshoot 1`] = `
<div>
  <section>
    <h1
      class="custom inline"
      data-testid="test-title"
      style="background: teal;"
    >
      holoroco
    </h1>
    <div>
      <h2>
        Counter: 
        1
      </h2>
      <button>
        Add
      </button>
      <button>
        Reset
      </button>
      <button>
        Substract
      </button>
    </div>
  </section>
</div>
`;
